File: src/breakerbox/report/shadow.py
"""#151 — aggregate would-trip events from shadow-mode runs into a report.

Shadow mode records a `would_trip` event (per reason) instead of enforcing, so a report over a
directory of run logs answers: *what would have tripped this week, and how much spend would have
been prevented?* — the adoption unlock (VISION §3.3): see the enforcement before you enable it.
"""

from __future__ import annotations

import json
from collections import Counter
from pathlib import Path


def aggregate_shadow(report_dir: str | Path) -> dict:
    """Scan every `*.jsonl` run log under `report_dir` and summarise its would-trip events."""
    runs: dict[str, list[dict]] = {}
    for p in sorted(Path(report_dir).glob("*.jsonl")):
        for line in p.read_text().splitlines():
            line = line.strip()
            if not line:
                continue
            try:
                ev = json.loads(line)
            except json.JSONDecodeError:
                continue
            runs.setdefault(ev.get("run_id", p.stem), []).append(ev)

    by_reason: Counter = Counter()
    by_policy: Counter = Counter()
    prevented = 0
    top: list[dict] = []
    for run_id, evs in runs.items():
        would = [e for e in evs if e.get("type") == "would_trip"]
        if not would:
            continue
        first = min(would, key=lambda e: e.get("seq", 0))
        for e in would:
            by_reason[e.get("detail", {}).get("reason")] += 1
            by_policy[e.get("detail", {}).get("policy")] += 1
        # spend that happened AFTER the first would-trip is what enforcement would have prevented
        final = max((e.get("cumulative_microusd") or 0) for e in evs)
        run_prevented = max(0, final - (first.get("cumulative_microusd") or 0))
        prevented += run_prevented
        top.append({
            "run_id": run_id,
            "reason": first.get("detail", {}).get("reason"),
            "prevented_micro": run_prevented,
        })
    top.sort(key=lambda r: r["prevented_micro"], reverse=True)
    return {
        "runs_scanned": len(runs),
        "runs_would_trip": len(top),
        "by_reason": dict(by_reason),
        "by_policy": dict(by_policy),
        "would_prevent_micro": prevented,
        "would_prevent_usd": round(prevented / 1_000_000, 6),
        "top_runs": top[:10],
    }


def render_shadow(summary: dict) -> str:
    """Human-readable one-screen summary of a shadow report."""
    lines = [
        f"Shadow report — {summary['runs_would_trip']}/{summary['runs_scanned']} "
        "run(s) would have tripped",
        f"  would have prevented: ${summary['would_prevent_usd']:.4f}",
    ]
    if summary["by_reason"]:
        lines.append("  by reason:")
        for reason, n in sorted(summary["by_reason"].items(), key=lambda kv: -kv[1]):
            lines.append(f"    {reason}: {n}")
    return "\n".join(lines)

File: cloud/dashboard/lib/shadow.ts
// #154b — hosted shadow report aggregation.
//
// SOURCE OF TRUTH: src/breakerbox/report/shadow.py `aggregate_shadow`. Keep this in lockstep — the
// Python version aggregates local *.jsonl run logs; this one aggregates the same events out of the
// cloud `events` table. Semantics mirrored exactly:
//   - a run "would trip" if it has any `would_trip` event (guard() dedups them per reason);
//   - `first` would-trip = the one with the smallest seq;
//   - a run's prevented spend = max(cumulative across ALL its events) − first would-trip's cumulative
//     (clamped ≥ 0): the money that was spent AFTER the point enforcement would have stopped it;
//   - by_reason / by_policy count every would_trip event's decision.reason / .policy.
// lib/shadow.test.ts locks this against hand-computed fixtures (multi-reason, single, clean run).

export type ShadowEvent = {
  run_id: string;
  seq: number | null;
  type: string;
  cumulative_microusd: number | null;
  detail: Record<string, unknown> | null;
};

export type ShadowSummary = {
  runs_scanned: number;
  runs_would_trip: number;
  by_reason: Record<string, number>;
  by_policy: Record<string, number>;
  would_prevent_micro: number;
  top_runs: { run_id: string; reason: string; prevented_micro: number }[];
};

const str = (v: unknown, fallback = "unknown") => (typeof v === "string" ? v : fallback);

export function aggregateShadow(events: ShadowEvent[]): ShadowSummary {
  const runs = new Map<string, ShadowEvent[]>();
  for (const e of events) {
    const arr = runs.get(e.run_id);
    if (arr) arr.push(e);
    else runs.set(e.run_id, [e]);
  }

  const by_reason: Record<string, number> = {};
  const by_policy: Record<string, number> = {};
  let prevented = 0;
  const top: { run_id: string; reason: string; prevented_micro: number }[] = [];

  for (const [run_id, evs] of runs) {
    const would = evs.filter((e) => e.type === "would_trip");
    if (would.length === 0) continue;
    const first = would.reduce((a, b) => ((a.seq ?? 0) <= (b.seq ?? 0) ? a : b));
    for (const e of would) {
      const reason = str(e.detail?.reason);
      const policy = str(e.detail?.policy);
      by_reason[reason] = (by_reason[reason] ?? 0) + 1;
      by_policy[policy] = (by_policy[policy] ?? 0) + 1;
    }
    const final = Math.max(...evs.map((e) => e.cumulative_microusd ?? 0));
    const runPrevented = Math.max(0, final - (first.cumulative_microusd ?? 0));
    prevented += runPrevented;
    top.push({ run_id, reason: str(first.detail?.reason), prevented_micro: runPrevented });
  }
  top.sort((a, b) => b.prevented_micro - a.prevented_micro);

  return {
    runs_scanned: runs.size,
    runs_would_trip: top.length,
    by_reason,
    by_policy,
    would_prevent_micro: prevented,
    top_runs: top.slice(0, 10),
  };
}
